### PR TITLE
Update E3SM config_compilers.xml to V2

### PR DESCRIPTION
**Pull request by jgfouca**
Update E3SM config_compilers.xml to v2 format.

The v2 format is more configurable and can output cmake as well as make output. v2 is also more testable and has more tests in the CIME regression test suite.

Also, remove PFLOTRAN hacks from root compiler block. There are commented-out lines in the cades compiler block that can be used to enable pflotran.

This will make it so all CIME-enabled models (e3sm, cesm) are using the same format for compiler settings, making it more likely that mistakes will be noticed and caught.  V1 was not even working correctly on melvin and was neglecting to add some of the debug flags.

[BFB]

### CODE REVIEW

**Comment by jgfouca** [comment]
Hi reviewers,

This is a high-impact PR and will therefore need a close review. The only obvious verification technique I can think of is to build a case with and without this change and visually inspect the compilation lines, looking at the flags. Unfortunately, this won't necessarily be enough since I just discovered that V1 was not even working properly, at least on melvin, it was not providing all the debug flags that it should have been.

**Comment by rljacob** [comment]
Can you add to the description some motivation for updating to v2?

**Comment by fmyuan** [comment]
@jgfouca - it's not suggested to remove the PFLOTRAN relevant options in compiler root block. That's totally different from what in machine files. The former is for adding '-D???' cppdef to compiler/linker options. The latter are user-options to switch those cppdef. 
- Fengming

**Comment by jgfouca** [comment]
@rljacob , will do.

@fmyuan , I have no choice. Those attributes don't work in V2.

**Comment by rljacob** [comment]
@fmyuan that block was identified as a hack.  Time to implement it correctly.  See what was done for PETSC, TRILLIONS etc in cime/scripts/Tools/Makefile.

File: cime/config/e3sm/machines/config_compilers.xml [hunk 1]
```diff
@@ -78,24 +77,147 @@ for mct, etc.
 <!-- Define default values that can be overridden by specific
      compilers -->
 <compiler>
+  <CPPDEFS>
+    <append MODEL="pop"> -D_USE_FLOW_CONTROL </append>
+  </CPPDEFS>
```
Comment (rljacob) [inline cime/config/e3sm/machines/config_compilers.xml]: Delete these lines.  We don't support MODEL=pop anymore.

**Comment by jgfouca** [comment]
@rljacob , I removed all "pop" references from the file.

**Comment by fmyuan** [comment]
@rljacob @jgfouca  - Then, after you're done with the changes, we will figure out how to deal with this following protocols defined.

**Comment by jgfouca** [comment]
Merged to next, will be interesting to see what the dashboard looks like tomorrow.

**Comment by worleyph** [comment]
So you are not waiting for all of the reviews?

**Comment by jgfouca** [comment]
@worleyph , i will try to get as many reviews as possible before merging to master.

**Comment by worleyph** [comment]
I'll summarize when I submit my review, but so far:

Titan/pgi: 

     -compset A_WCYCL1950S_CMIP6_LR -res ne30_oECv3_ICG

log output are bit-for-bit with master with a typical layout (with the exception of minor differences in timing and memory information, which vary from run to run). 

Looking at compiler flags for a few files in atm and lnd and ocean, they are the same, though the order is different for a few, and white space separator also differs for a few. (Would be nice if these were identical - would make it easier to compare.) Differences I am seeing include placement of -mp and -DLINUX

**Comment by worleyph** [comment]
@jgfouca , found a problem. For Titan/intel, compilation is dying when building gptl (first component it tries):

     ...
     c -I/autofs/nccs-svm1_home1/worley/E3SM/jgfouca/cime/update_compilers_to_v2/E3SM/cime/src/share/timing  \
     -O2 -fp-model precise -std=gnu99   -openmp   -O2 -debug minimal   -O2    -DLINUX  \ 
     -DHAVE_NANOTIME -DBIT64 -DHAVE_VPRINTF \
     -DHAVE_BACKTRACE -DHAVE_SLASHPROC -DHAVE_COMM_F2C -DHAVE_TIMES  \
     -DHAVE_GETTIMEOFDAY    -DFORTRANUNDERSCORE -DNO_R16 -DCPRINTEL -DHAVE_MPI \
     /autofs/nccs-svm1_home1/worley/E3SM/jgfouca/cime/update_compilers_to_v2/E3SM/cime/src/share/timing/gptl.c
     ...
     gmake: c: Command not found
     gmake: [gptl.o] Error 127 (ignored)

**Comment by rljacob** [comment]
@jgfouca what kind of tests have you done with the branch on how many machines?

**Comment by jgfouca** [comment]
@rljacob , building on melvin, sandiatoss3, anlworkstation

**Comment by jgfouca** [comment]
@worleyph , thanks, i will look at titan/intel.

**Comment by worleyph** [comment]
@jgfouca , same error occurs with titan/pgiacc .

**Comment by jgfouca** [comment]
@worleyph , yes, there's something way wrong with the Macros.make files being produced.

**Comment by jgfouca** [comment]
@worleyph , OK, titan + intel/pgiacc appear to be working again.

**Comment by worleyph** [comment]
@jgfouca , thanks. I have another obligation this afternoon, but will try these out this evening (I hope).

**Comment by jgfouca** [comment]
all, please hold off on reviewing this PR. Bugs were discovered and I'll need to do another CIME update before this is ready.

**Comment by jgfouca** [comment]
Rebased to latest master. Merging back to next.

**Comment by rljacob** [comment]
@jgfouca you may want to try Titan yourself in case @worleyph can't get to it.

File: cime/config/e3sm/machines/config_compilers.xml [hunk 2]
```diff
@@ -112,1456 +230,2059 @@ for mct, etc.
...
+    <append> -DLINUX </append>
+  </CPPDEFS>
+  <FFLAGS>
+    <append DEBUG="TRUE"> -g -traceback  -O0 -fpe0 -check  all -check noarg_temp_created -ftrapuv </append>
+  </FFLAGS>
+  <NETCDF_PATH> $ENV{NETCDF_HOME}</NETCDF_PATH>
+  <PIO_FILESYSTEM_HINTS>lustre</PIO_FILESYSTEM_HINTS>
+  <PNETCDF_PATH>$ENV{PNETCDFROOT}</PNETCDF_PATH>
+  <SLIBS>
+    <base> -L$NETCDF_PATH/lib -lnetc
```
Comment (amametjanov) [inline cime/config/e3sm/machines/config_compilers.xml]: FFLAGS in debug mode are not getting added to Macros.make for theta+intel.
* Before:
```
ifeq ($(DEBUG), TRUE)
   FFLAGS +=  -O0 -g -check uninit -check bounds -check pointers -fpe0 -check noarg_temp_created
   CFLAGS +=  -O0 -g
endif
```
* After:
```
ifeq ($(COMPILER),intel)
  SLIBS := $(SLIBS)  -L$(NETCDF_DIR)/lib -lnetcdff -L$(NETCDF_DIR)/lib -lnetcdf -Wl,-rpath -Wl,$(NETCDF_DIR)/lib
  SLIBS := $(SLIBS)  -mkl -lpthread
  CPPDEFS := $(CPPDEFS)  -DFORTRANUNDERSCORE -DNO_R16 -DCPRINTEL
  CPPDEFS := $(CPPDEFS)  -DARCH_MIC_KNL
  ifeq ($(compile_threaded),true)
    CFLAGS := $(CFLAGS)  -openmp
    CFLAGS := $(CFLAGS)  -qopenmp
    FFLAGS := $(FFLAGS)  -qopenmp
  endif
  ifeq ($(DEBUG),FALSE)
    CFLAGS := $(CFLAGS)  -O2 -debug minimal
    FFLAGS := $(FFLAGS)  -O2 -debug minimal -qno-opt-dynamic-align -fp-speculation=off
  endif
  ifeq ($(DEBUG),TRUE)
    CFLAGS := $(CFLAGS)  -O0 -g
  endif
  ifeq ($(compile_threaded),true)
    FFLAGS_NOOPT := $(FFLAGS_NOOPT)  -openmp
    FFLAGS_NOOPT := $(FFLAGS_NOOPT)  -qopenmp
    LDFLAGS := $(LDFLAGS)  -openmp
    LDFLAGS := $(LDFLAGS)  -qopenmp
  endif
endif
```

Comment (jgfouca) [reply cime/config/e3sm/machines/config_compilers.xml]: @amametjanov , thanks, I will investigate.

**Comment by jgfouca** [comment]
@amametjanov @rljacob @worleyph , with the latest CIME update, the next branch appears to be working correctly with this PR.

**Comment by rljacob** [comment]
There are still problems on cori-knl.  @ndkeen has details.

File: cime/config/e3sm/machines/config_compilers.xml (hunk 2 above)
Comment (amametjanov) [reply cime/config/e3sm/machines/config_compilers.xml]: Debug `FFLAGS` are still missing on the latest `next` branch.

**Comment by jgfouca** [comment]
I think the cori-knl problems have been fixed.

**Comment by jgfouca** [comment]
@amametjanov , when I look at theta, it appears to have debug flags set. Can you retry?

**Comment by amametjanov** [comment]
I can't find them:
```
azamatm@thetalogin1:~/repos/E3SM> git branch
  master
* next
azamatm@thetalogin1:~/repos/E3SM> git describe
v1.0.0-358-gc55a42b92
azamatm@thetalogin1:~/repos/E3SM> ./cime/scripts/create_test SMS_D_Ln5.ne4_ne4.FC5AV1C-L --no-build
azamatm@thetalogin1:~/repos/E3SM> grep -i "check uninit" /projects/ClimateEnergy_2/azamatm/SMS_D_Ln5.ne4_ne4.FC5AV1C-L.theta_intel.20180710_174915_doc1ht/Macros.make
azamatm@thetalogin1:~/repos/E3SM>
```

**Comment by jgfouca** [comment]
@amametjanov , OK, I've confirmed those flags now exist with the lastest CIME update now merged into this PR.

**Comment by jgfouca** [comment]
The next branch dashboard results look quite good. I've confirmed that the build fails on titan are not related to this PR, so I am going to merge.

**Pull request merged**